src/fu-release.c

```diff
@@ -769,22 +769,44 @@ fu_release_get_priority(FuRelease *self)
 	return self->priority;
 }
 
-static void
+/**
+ * fu_release_ensure_device_by_checksum:
+ * @self: a #FuRelease
+ * @component: (not nullable): a #XbNode
+ * @rel: (not nullable): a #XbNode
+ *
+ * Ensures any properties using the device checksum. These should only be attempted after the
+ * release requirements have been verified.
+ *
+ * NOTE: This will only populate properties from trusted metadata, and only for devices with the
+ * `md-only-checksum` private flag set.
+ **/
+void
 fu_release_ensure_device_by_checksum(FuRelease *self, XbNode *component, XbNode *rel)
 {
 	g_autoptr(GPtrArray) device_checksums = NULL;
 
+	g_return_if_fail(FU_IS_RELEASE(self));
+	g_return_if_fail(XB_IS_NODE(component));
+	g_return_if_fail(XB_IS_NODE(rel));
+
 	/* sanity check */
+	if (self->device == NULL)
+		return;
 	if (fu_device_get_checksums(self->device)->len == 0)
 		return;
+	if (!fu_device_has_private_flag(self->device, FU_DEVICE_PRIVATE_FLAG_MD_ONLY_CHECKSUM))
+		return;
+	if (!fu_release_has_flag(self, FWUPD_RELEASE_FLAG_TRUSTED_METADATA))
+		return;
```

Comment:
this is the one more common to fail for common case - I would put it at the start of sanity checks for speed purposes

Reply:
I agree it's a bit confusing, the function should be in the same place as fu_device_ensure_from_component() to reduce scope even more. I've refactored things to be a bit easier to understand.